dired-subtree.el

```diff
@@ -463,12 +463,7 @@ children."
 
 Return a string suitable for insertion in `dired' buffer."
   (with-temp-buffer
-    (let ((insert-dir-fun  (if (and (featurep 'tramp)
-                                    (tramp-tramp-file-p dir-name)
-                                    (tramp-sh-handle-file-directory-p dir-name))
-                               #'tramp-handle-insert-directory
-                             #'insert-directory)))
-      (funcall insert-dir-fun dir-name dired-listing-switches nil t))
+    (insert-directory dir-name dired-listing-switches nil t)
```

Comment:
If you look at dc631bd1ab02b1e1b257d765d03d30edd40cf2a9 it is pretty much exact reverse of your change here.  The comment says it fixed issue #22.  So I'm not sure what's going on (it was a long time ago :blush:)

Reply:
Actually, it looks like it was you who proposed that patch :D

Reply:
Yes I know! I was browsing the issues list and remembered that I did submit a patch about this issue which is why I thought I would go ahead and try to figure out what the problem was since when I submitted the patch I was just starting out with Emacs.

Anyway, the problem is that `dired-get-filename` does not add a final `/` when getting a directory name. I dug through the code of `tramp-sh-handle-insert-directory` and found that it always appends a `.` to the file name it is provided when asked for a full directory listing without passing the wildcard argument to `insert-directory` (`tramp-sh-handle-insert-directory` will eventually get called by `insert-directory` for shell like remote connections thanks to [magic file names](https://www.gnu.org/software/emacs/manual/html_node/elisp/Magic-File-Names.html)). This means that if your provide a file name `foo/bar`, the file name it will use when asking for the listing from the remote system is `foo/bar.`, but this isn't a problem when the file name passed to it is a directory name, i.e. `foo/bar/`. You can see this if you search for `tramp-send-command` in `tramp-sh-handle-insert-directory`. It is the first call to that function. I should also mention I'm on Emacs 26.1, but the same code exists in Emacs 24 so this is the same issue that I addressed in dc631bd.

dc631bd fixed the issue for me back then because `tramp-handle-insert-directory` uses a different method compared to `tramp-sh-handle-insert-directory` to get the listing. At least on my system it goes through `ls-lisp--insert-directory` which advises `insert-directory` and when called on a remote directory will defer to its own method of obtaining the listing which does not have the same file name issue above.

Reverting back to the behavior before dc631bd while ensuring that the file name passed to `insert-directory` is a directory name makes `dired-subtree-insert` more general for different kinds of remote connections because it avoids the use of `tramp-sh-handle-file-directory-p` which is specific to shell like remote connections and won't work for others. Also, ensuring that the file name passed to `insert-directory` is a directory file name fixes the issue originally addressed in dc631bd due to the above explanation.

Lots and lots of code archaeology...

Reply:
So all the magic is essentially hidden in the call to `file-name-as-directory` which will sanitize the input for the magic handler.

My gods what a mess.  Seems like a guest post for http://emacshorrors.com/ would be in order :D